### DISCUSSION TITLE
Dependency Management: Import CMake projects?

### DISCUSSION THREAD

**Discussion by Quincunx271** [vector-of-bool/dds#83] [Ideas]
I know this goes against the philosophy of DDS, but the reality is that there are a lot of dependencies out there, and it's a lot of work to write a DDS port / porting script for every library we want to use. Furthermore, sometimes the project does some complex things with the build system that makes it difficult to see how to translate it to the structure DDS requires. Understanding this, what if DDS was able to consume projects with a CMake build system?

What I mean by this is that we could have DDS invoke CMake to build the dependencies, extracting information from specified CMake targets. The user would still have to add a DDS repo pointing to the CMake project, but it could contain some metadata that tells DDS how to invoke CMake to build the project with consistent flags.

Although this is a complex solution to the problem, the benefit is that it would allow the easy consumption of most libraries even if they were not ported to the DDS style.

**Comment by vector-of-bool** [comment]
I think there's a few ways to approach the task:

1. Have `dds` directly download and import the CMake project from its original upstream source.
2. Support a more easy mechanical transformation of a CMake project into a consumable source distribution which `dds` can import  and build.

Option (1) will definitely require changes to `dds` itself, whereas option (2) could be approached with a supplemental tool. In the [dds-ports](https://github.com/vector-of-bool/dds-ports/) project, it leans toward option (2) and has a lot of Python utilities that will patch-up source trees into source distributions.

I'm also inclined toward option (2) in particular because I have thought about making the source-dist format that `dds` uses into a tool-agnostic packaging format. That is, it would be handy if CMake, Meson, Conan, etc. could just "do the right thing" when it sees a source distribution tarball, and the sdist itself will have no (or minimal) dds-specifics inside of it.

For option (1), these are the following points to consider:

1. Packaging of sources for downstream distribution. The sdist format is extremely minimal, and since CMake can execute arbitrary code at configure-time, it would mean that it is unsafe to make any assumptions about what can/cannot be included within the sdist. Alternatively, don't do any packaging, and just give a project a way to import directly from a directory/zip/tgz/repo; this was how it was done in earlier `dds` versions for its own packages, and it worked but was somewhat fragile.
2. Because CMake executes arbitrary code at configure-time, this presents enormous security implications on automatically downloading and building untrusted code as part of a `dds build` operation.
3. The `dds build` will need to know if/how/when to propagate toolchain options down into CMake projects.
4. CMake's File API will be a useful mechanism to obtain details such as `#include` search paths and files-to-compile.
5. Packages would be consumable with no/minimal changes to their original source tree.

For option (2), there are these points:

1. `dds` would not need to be changed to understand these packages.
2. A supplemental tool could be created to generate sdist tarballs from CMake projects. This tool might be integrated into `dds` itself as a new subcommand (i.e. `dds pkg create --cmake`).
3. For many (especially modern) CMake projects, CMake isn't even in the picture and we can simply shuffle sources around to create an sdist. This is the case for most packages imported by `dds-ports`.
4. If a project performs code generation at configure or build time, those generated sources would simply need to be included in the sdist as regular source files.
5. If the project performs code generation that encodes toolchain-specific or platform-specific information, that codegen needs to be *removed and rewritten*. For example, the `libsodium` project is imported by `dds-ports`, but the upstream project contains configure-time platform feature detection (such as vectorization intrinsics). `dds-ports` replaces that configure-time feature detection with a "smart `config.h`" header that performs similar feature detection (refer: https://github.com/vector-of-bool/dds-ports/blob/master/ports/libsodium_ports.py). This relies heavily on `__has_include()`, which is supported by the major compilers for both C and C++ code.
6. Any of these mechanical transformations can be added to `dds-ports` to make them globally available, and such tooling can be offered as part of `dds` itself.

My inclination is towards an integrated tool within `dds` that can create sdist-archives from CMake projects, and which generated archives are able to be compiled by `dds`. A few ideas on how it might work:

1. Create an empty directory `<sdist-dir>`
2. Run CMake to configure a project at lies within the given path. Enable the File API to request project metadata.
3. Find every `#include` search path for the project that is a subdirectory of the project's source or build directories. Recursively copy every header file from each found search path, relative to that search path as `<search-dir>/<subpath>` into `<sdist-dir>/src/<subpath>`. If any two files collide, issue a diagnostic and abort.
4. Copy every compiled source file into `<sdist-dir>/src`. This may be somewhat fragile, because source files may assume some relative `#include` directive will work, so the structure of the original source tree should be maintained as well as possible. If any two files collide, issue a diagnostic and abort.
5. Remove any source files that are used solely for the generation of executables, as the sdist is only concerned with generated libraries. This will automatically prune tests, benchmarks, examples, and utilities from the sdist. (Shipping of tests may be desired in the future, but is very difficult to find these mechanically without deep introspection.)
6. If the project defines some preprocessor directives globally to all compiled source files, ~weep~ a configuration header could be generated into the `<sdist-dir>/src/` tree, and each compiled file can have an `#include` directive inserted that will include this file. This will allow each compiled file to see the definitions that it expected to see as part of the CMake project. This may still be somewhat fragile. 
7. If the project defines any other compiler flags that are not specifically allowed (e.g. warning flags, `-fPIC`, `/bigobj`), issue a diagnostic. Maybe abort? Maybe continue?
8. Attempt to build the resulting sdist with `dds`. If this succeeds, you may have successfully created a source distribution.
9. As an additional layer of verification, we can perform the above steps _again_, but specify a CMake toolchain file and/or `CMAKE_BUILD_TYPE`. If the resulting sdist directory is not byte-wise identical to the other, then it means the project _may_ have encoded toolchain/platform information into the generated tree, and it is not safe to ship any particular generated sdist.

**Reply by Quincunx271** [reply]
I am referring to the method 1 when I say, "Import CMake projects." I do indeed prefer the method 2&mdash;simple sdist with a tool to automate packaging a CMake project&mdash;that you suggest, but I have 2 reservations:

 1. This requires the "user" to do package maintenance style of work. If I want to use a library, I have to figure out how to convert it to a sdist. If this is as easy as pointing a tool at it, it's fine, but if the tool doesn't work because the project does something weird, suddenly I have to do a lot more work to figure out how to adapt the project to something that DDS supports.
 2. Many CMake projects are not structured in a way that DDS supports. E.g. LLVM, freetype, etc. It may be possible to adapt the project to DDS's style, but it can be quite difficult. It will be easier to use DDS if this adaptation can be postponed for later, with direct consumption via CMake to start things out.

I suppose what I'm trying to say with 1. is that as a user, I want to depend on `libgui` and `libaudio` and `libparser`. I don't want to have to go through the work of saying, "Here's how you can adapt `libgui` to be compiled with DDS" when that work can be very difficult: diving through complicated CMakeLists.txt to parse through conditional configuration so that I can translate that to preprocessor directives, making sense of odd file naming conventions for the library where some source files shouldn't be compiled directly, etc. I just want to say, "I have a dependency on `libgui`. Oh, don't worry, I know it doesn't support DDS, but it does support CMake with these knobs that you need to know about."

What I'm trying to say with 2. is that it's hard to use DDS when certain valuable libraries are locked away because nobody has yet done the work to port them to DDS or because DDS cannot currently support them. A workaround that allows consuming any library that already supports CMake would mean that I can use those libraries without having to wait for DDS support.

I absolutely agree that the simple sdist is preferable. However, I think it would be useful to have a "cmake-sdist" so that DDS users can use libraries before the libraries support DDS or DDS supports the libraries. The understanding should always be that a "cmake-sdist" package is a workaround which should be replaced in the future when someone gets around to doing the work to do a proper DDS port.

**Reply by Quincunx271** [reply]
Should the dds-ports python package be uploaded to PYPI?

**Reply by Quincunx271** [reply]
Another idea came up which could resolve this. In particular, `dds` could be kept aligned with its philosophy and just work with simple source distributions, but a `meta-dds`&mdash;probably living in a separate repository&mdash;can exist to support CMake / other source distributions. This `meta-dds` can do two things:

 1. Work with "meta-sdists" which don't represent a source distribution on its own, but must have some processing with the toolchain information to produce a source distribution. By invoking the build through `meta-dds`, it can feed the toolchain information into the meta-sdist and run `cmake` to do whatever it needs to.
    - This opens more ideas, such as possibly abstracting the interface into some sort of "toolchain/platform specification format" that could be used for CMake, Meson, etc.
    - There is also the question of whether `meta-dds` should have CMake run the build or whether it should instead only have CMake run its configuration + any sourcefile generation (specified via CMake script, perhaps) to create a simple sdist that can be handed to `dds` to actually build. Generating a simple sdist seems like the better approach, but idk if it's generally possible.
    - There is also the question of how `meta-dds` can pass dependencies into the CMake project. Ideally, if the CMake project had a `find_package(SomeDependency ...)`, that should use the same dependency as if the package.json specified `some-dependency` in its `uses`/`depends`, so something would have to exist to work out this issue.
 2. Do any initial work on the project itself to create a sdist that `dds` proper can build. E.g. run `llvm-tblgen`, run `bison` / `flex`, etc.

This could be a reasonable solution to the problem, and I'm currently most fond of this idea, but it does have the disadvantage of being yet another tool, like `cmake` being a build system generator rather than a build system on its own.

**Comment by Quincunx271** [comment]
Stepping back and looking more broadly on why I want this: I've found that the most successful tools usually are those which are either backwards compatible with or easily integrated with the existing infrastructure.

Positive examples:
 + C++ being "built on" C
 + JVM languages being able to leverage Java libraries, e.g. Scala, Kotlin, Groovy, etc.

Counter examples:
 - Python 3's breaking changes with Python 2 caused issues when trying to use dependencies which depended on Python 2 behavior.

IMO, if we want DDS to be broadly successful, it should be able to trivially consume CMake packages, because that gives compatibility with the current C++ ecosystem. If it really is true that very few projects cannot be adapted by the proposed tool, then that meets the "trivially consumable" requirement so it _should_ be okay.

However, I feel that it's not so simple, because those very few projects which cannot be automatically adapted are likely to be large, complicated projects which are hard requirements for some classes of users, e.g. LLVM, Qt, etc. It is my impression that there are enough of these projects that it is infeasible to provide them all in a central DDS repo.
